src/decimal.rs

```diff
@@ -514,116 +528,168 @@ impl Decimal {
-                    break;
-                } else {
-                    div_by_u32(&mut offset, POWERS_10[9]);
-                    // Only 9 as this array starts with 1
-                    diff -= 9;
-                }
-            }
-            let mut diff = old_scale - dp;
-            while diff > 0 {
-                if diff < 10 {
-                    mul_by_u32(&mut offset, POWERS_10[diff as usize]);
-                    break;
-                } else {
-                    mul_by_u32(&mut offset, POWERS_10[9]);
-                    // Only 9 as this array starts with 1
-                    diff -= 9;
-                }
+        // Do some midpoint rounding checks
+        // We're actually doing two things here.
+        //  1. Figuring out midpoint rounding when we're right on the boundary. e.g. 2.50000
+        //  2. Figuring out whether to add one or not e.g. 2.51
+        // For this, we need to figure out the fractional portion that is additional to
+        // the rounded number. e.g. for 0.12345 rounding to 2dp we'd want 345.
+        // We're doing the equivalent of losing precision (e.g. to get 0.12)
+        // then increasing the precision back up to 0.12000
+        let mut offset = [self.lo, self.mid, self.hi];
+        let mut diff = old_scale - dp;
+
+        while diff > 0 {
+            if diff < 10 {
+                div_by_u32(&mut offset, POWERS_10[diff as usize]);
+                break;
+            } else {
+                div_by_u32(&mut offset, POWERS_10[9]);
+                // Only 9 as this array starts with 1
+                diff -= 9;
             }
+        }
 
-            let mut decimal_portion = [self.lo, self.mid, self.hi];
-            sub_internal(&mut decimal_portion, &offset);
+        let mut diff = old_scale - dp;
 
-            // If the decimal_portion is zero then we round based on the other data
-            let mut cap = [5, 0, 0];
-            for _ in 0..(old_scale - dp - 1) {
-                mul_by_u32(&mut cap, 10);
+        while diff > 0 {
+            if diff < 10 {
+                mul_by_u32(&mut offset, POWERS_10[diff as usize]);
+                break;
+            } else {
+                mul_by_u32(&mut offset, POWERS_10[9]);
+                // Only 9 as this array starts with 1
+                diff -= 9;
             }
-            let order = cmp_internal(&decimal_portion, &cap);
-            match order {
-                Ordering::Equal => {
-                    if (value[0] & 1) == 1 {
+        }
+
+        let mut decimal_portion = [self.lo, self.mid, self.hi];
+        sub_internal(&mut decimal_portion, &offset);
+
+        // If the decimal_portion is zero then we round based on the other data
+        let mut cap = [5, 0, 0];
+        for _ in 0..(old_scale - dp - 1) {
+            mul_by_u32(&mut cap, 10);
+        }
+        let order = cmp_internal(&decimal_portion, &cap);
+
+
+        match strategy {
```

Comment:
I'm thinking about trying to refactor strategies into a trait, that way this crate can provide some default implementations, but users can extend it further if they wish to do so.

If this approach is taken, I think some additional methods would need to be exposed to make it possible for alternative strategies to be implemented.

Reply:
Interesting idea. I like the extensibility aspect of this - how would you ensure it remained ergonomic?

Reply:
Assuming you're talking about ergonomics for the method and/or trait, my thought was that we could have this method call out to a method defined on the trait e.g. `round_to_dp(n: u32)`, from there it's up to the strategy on _how_ to do the rounding.

This should give the solution the most flexibility, as the implementation won't know anything about the rounding, except that it returns a `Decimal`.

In summary, something like:
```rust
trait RoundingStrategy {...}

pub fn round_dp_with_strategy(&self, dp: u32, strategy: RoundingStrategy) -> Decimal {
  strategy.round_to_dp(self, dp)
}
```

Can you see any problems and/or concerns with that approach?

Reply:
Playing it out from an integration perspective; are you thinking:
```
let number = dec!(1.2345);
let _ = number.round_with_strategy(2, BankersRoundingStrategy::default());
```

You could also feasibly invert this relationship also:
```
let number = dec!(1.2345);
let bankers_rounding = BankersRounding::with_dp(2);
let _ = bankers_rounding.round(number);
```

The problem I see with each is accessing/setting the internals effectively for efficiency. I think that at the moment, an `enum` is great for optimizing towards efficiency because it allows everything to be handled within the same scope. If another rounding strategy is required later on then it is feasible for someone to do this using the public API of the decimal type in an approach similar to the second example (at a small perf cost).

On a side note, the function name is starting to get quite long and since it'll become part of the public API is the perfect time to think about it. I've been trying to think of alternatives to assist with this but the best I can come up with is _maybe_ dropping the `_dp`. I'm still a little torn here... I'm definitely open to keeping it the same but if you think of something more succinct then I'm all ears!

Reply:
That's along the lines of what I was thinking of, however another alternative would be to set the rounding strategy when the decimal is created. `BankersRounding` would be the default, then the user could update it or use a specialised constructor like:

```rust
let number = Decimal::new_with_strategy(134, 2, Rounding::RoundHalfUp);
let rounded = number.round_dp(2);
```

This would eliminate the naming problem entirely 🙂 

I can definitely see where you're coming from, it'd be a trade-off of performance for flexibility, I'm just not sure what that impact would be, and whether it'd be worth it in the long run.